data-prepper-plugins/grok-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/grok/GrokPrepperIT.java

```diff
@@ -0,0 +1,286 @@
+
+        List<Record<String>> grokkedRecords = (List<Record<String>>) grokPrepper.doExecute(Collections.singletonList(record));
+
+        assertThat(grokkedRecords.size(), equalTo(1));
+        assertThat(grokkedRecords.get(0), notNullValue());
+        assertThat(equalRecords(grokkedRecords.get(0), resultRecord), equalTo(true));
+    }
+
+    @Test
+    public void testSingleMatchMultiplePatternWithDefaults() throws JsonProcessingException {
+        final Map<String, List<String>> matchConfig = new HashMap<>();
+        final List<String> patternsToMatchMessage = new ArrayList<>();
+        patternsToMatchMessage.add("%{COMMONAPACHELOG}");
+        patternsToMatchMessage.add("%{IPORHOST:custom_client_field}");
+
+        matchConfig.put("message", patternsToMatchMessage);
+
+        pluginSetting.getSettings().put(GrokPrepperConfig.MATCH, matchConfig);
+        grokPrepper = new GrokPrepper(pluginSetting);
+
+        String testData = "{\"message\":\"127.0.0.1 user-identifier frank [10/Oct/2000:13:55:36 -0700] \\\"GET /apache_pb.gif HTTP/1.0\\\" 200 2326\"}";
+        Record<String> record = new Record<>(testData);
+
+        String resultData = "{\"message\":\"127.0.0.1 user-identifier frank [10/Oct/2000:13:55:36 -0700] \\\"GET /apache_pb.gif HTTP/1.0\\\" 200 2326\","
+                .concat("\"clientip\":\"127.0.0.1\",")
+                .concat("\"ident\":\"user-identifier\",")
+                .concat("\"auth\":\"frank\",")
+                .concat("\"timestamp\":\"10/Oct/2000:13:55:36 -0700\",")
+                .concat("\"verb\":\"GET\",")
+                .concat("\"request\":\"/apache_pb.gif\",")
+                .concat("\"httpversion\":\"1.0\",")
+                .concat("\"response\":\"200\",")
+                .concat("\"custom_client_field\":\"127.0.0.1\",")
+                .concat("\"bytes\":\"2326\"}");
+
+        Record<String> resultRecord = new Record<>(resultData);
+
+        List<Record<String>> grokkedRecords = (List<Record<String>>) grokPrepper.doExecute(Collections.singletonList(record));
+
+        assertThat(grokkedRecords.size(), equalTo(1));
+        assertThat(grokkedRecords.get(0), notNullValue());
+        assertThat(equalRecords(grokkedRecords.get(0), resultRecord), equalTo(true));
+    }
+
+    @Test
+    public void testSingleMatchTypeConversionWithDefaults() throws JsonProcessingException {
+        final Map<String, List<String>> matchConfig = new HashMap<>();
+        matchConfig.put("message", Collections.singletonList("\"(?:%{WORD:verb} %{NOTSPACE:request}(?: HTTP/%{NUMBER:httpversion})?|%{DATA:rawrequest})\" %{NUMBER:response:int} (?:%{NUMBER:bytes:float}|-)"));
+
+        pluginSetting.getSettings().put(GrokPrepperConfig.MATCH, matchConfig);
+        grokPrepper = new GrokPrepper(pluginSetting);
+
+        String testData = "{\"message\":\"127.0.0.1 user-identifier frank [10/Oct/2000:13:55:36 -0700] \\\"GET /apache_pb.gif HTTP/1.0\\\" 200 2326\"}";
```

Comment:
nit-pick: It seems to me the testData does not include %{DATA:rawrequest}. From the context I infer it is to cover empty_match = False? If so, we should better comment on this field.

Reply:
Yeah rawrequest being null makes it easier to test keep_empty_captures and to test merging maps with null values to make sure no NPE's can occur. Can you elaborate on what you mean by "better comment on this field"?

Reply:
If I am correct, rawrequest key is not expected to appear in the keys of result data in this test cases while it appears in the grokking pattern. So we could call that out in the comment (maybe above result data)

Reply:
I will put a comment somewhere for this